File: campstation/backend/src/main/java/com/campstation/service/AuthService.java
package com.campstation.service;

import java.time.LocalDateTime;

import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.campstation.config.JwtUtil;
import com.campstation.dto.AuthResponse;
import com.campstation.dto.LoginRequest;
import com.campstation.dto.RefreshTokenRequest;
import com.campstation.dto.RegisterRequest;
import com.campstation.dto.UserDto;
import com.campstation.entity.RefreshToken;
import com.campstation.entity.User;
import com.campstation.repository.RefreshTokenRepository;
import com.campstation.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;
    private final AuthenticationManager authenticationManager;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        // 이메일 중복 체크
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("이미 사용 중인 이메일입니다.");
        }

        // 사용자 생성
        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(User.Role.USER)
                .build();

        user = userRepository.save(user);

        // 토큰 생성
        String token = jwtUtil.generateToken(user.getEmail());
        String refreshToken = jwtUtil.generateRefreshToken(user.getEmail());

        // Refresh token 저장
        saveRefreshToken(user, refreshToken);

        return AuthResponse.builder()
                .token(token)
                .refreshToken(refreshToken)
                .user(UserDto.fromEntity(user))
                .build();
    }

    @Transactional
    public AuthResponse login(LoginRequest request) {
        // 인증
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getEmail(),
                        request.getPassword()
                )
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);

        // 사용자 조회
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("사용자를 찾을 수 없습니다."));

        // 토큰 생성
        String token = jwtUtil.generateToken(user.getEmail());
        String refreshToken = jwtUtil.generateRefreshToken(user.getEmail());

        // 기존 Refresh token 삭제 후 새로 저장
        refreshTokenRepository.deleteByUser(user);
        saveRefreshToken(user, refreshToken);

        return AuthResponse.builder()
                .token(token)
                .refreshToken(refreshToken)
                .user(UserDto.fromEntity(user))
                .build();
    }

    @Transactional
    public AuthResponse refreshToken(RefreshTokenRequest request) {
        String refreshTokenValue = request.getRefreshToken();

        // Refresh token 검증
        if (!jwtUtil.validateToken(refreshTokenValue)) {
            throw new RuntimeException("유효하지 않은 Refresh Token입니다.");
        }

        // Refresh token 조회
        RefreshToken refreshToken = refreshTokenRepository.findByToken(refreshTokenValue)
                .orElseThrow(() -> new RuntimeException("Refresh Token을 찾을 수 없습니다."));

        // 만료 체크
        if (refreshToken.isExpired()) {
            refreshTokenRepository.delete(refreshToken);
            throw new RuntimeException("만료된 Refresh Token입니다.");
        }

        // 사용자 조회
        User user = refreshToken.getUser();
        String email = user.getEmail();

        // 새 토큰 생성
        String newToken = jwtUtil.generateToken(email);
        String newRefreshToken = jwtUtil.generateRefreshToken(email);

        // Refresh token 갱신
        refreshTokenRepository.delete(refreshToken);
        saveRefreshToken(user, newRefreshToken);

        return AuthResponse.builder()
                .token(newToken)
                .refreshToken(newRefreshToken)
                .user(UserDto.fromEntity(user))
                .build();
    }

    public UserDto getCurrentUser() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String email = authentication.getName();

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("사용자를 찾을 수 없습니다."));

        return UserDto.fromEntity(user);
    }

    @Transactional
    public void logout() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String email = authentication.getName();

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("사용자를 찾을 수 없습니다."));

        refreshTokenRepository.deleteByUser(user);
        SecurityContextHolder.clearContext();
    }

    private void saveRefreshToken(User user, String tokenValue) {
        RefreshToken refreshToken = RefreshToken.builder()
                .token(tokenValue)
                .user(user)
                .expiryDate(LocalDateTime.now().plusDays(7))
                .createdAt(LocalDateTime.now())
                .build();

        refreshTokenRepository.save(refreshToken);
    }
}

File: psmo-community/backend/src/main/kotlin/com/psmo/service/AuthService.kt
package com.psmo.service

import com.psmo.dto.*
import com.psmo.model.*
import com.psmo.security.JwtConfig
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.transactions.transaction
import org.mindrot.jbcrypt.BCrypt
import java.time.LocalDateTime

class AuthService(private val jwtConfig: JwtConfig) {

    fun register(request: RegisterRequest): AuthResponse {
        // 이메일 중복 체크
        val existingUser = transaction {
            Users.selectAll().where { Users.email eq request.email }.singleOrNull()
        }

        if (existingUser != null) {
            throw IllegalArgumentException("이미 사용 중인 이메일입니다.")
        }

        // 사용자 생성
        val userId = transaction {
            Users.insert {
                it[name] = request.name
                it[email] = request.email
                it[password] = BCrypt.hashpw(request.password, BCrypt.gensalt())
                it[role] = "USER"
            } get Users.id
        }

        // 토큰 생성
        val token = jwtConfig.generateToken(request.email)
        val refreshToken = jwtConfig.generateRefreshToken(request.email)

        // Refresh token 저장
        saveRefreshToken(userId.value, refreshToken)

        val user = transaction {
            Users.selectAll().where { Users.id eq userId }
                .single()
                .let {
                    UserDto(
                        id = it[Users.id].value,
                        name = it[Users.name],
                        email = it[Users.email],
                        role = it[Users.role]
                    )
                }
        }

        return AuthResponse(token, refreshToken, user)
    }

    fun login(request: LoginRequest): AuthResponse {
        // 사용자 조회
        val user = transaction {
            Users.selectAll().where { Users.email eq request.email }.singleOrNull()
        } ?: throw IllegalArgumentException("이메일 또는 비밀번호가 올바르지 않습니다.")

        // 비밀번호 확인
        if (!BCrypt.checkpw(request.password, user[Users.password])) {
            throw IllegalArgumentException("이메일 또는 비밀번호가 올바르지 않습니다.")
        }

        val userId = user[Users.id].value

        // 토큰 생성
        val token = jwtConfig.generateToken(request.email)
        val refreshToken = jwtConfig.generateRefreshToken(request.email)

        // 기존 Refresh token 삭제 후 새로 저장
        transaction {
            RefreshTokens.deleteWhere { RefreshTokens.userId eq userId }
        }
        saveRefreshToken(userId, refreshToken)

        val userDto = UserDto(
            id = user[Users.id].value,
            name = user[Users.name],
            email = user[Users.email],
            role = user[Users.role]
        )

        return AuthResponse(token, refreshToken, userDto)
    }

    fun refreshToken(request: RefreshTokenRequest): AuthResponse {
        val refreshTokenValue = request.refreshToken

        // Refresh token 조회
        val tokenData = transaction {
            RefreshTokens.selectAll().where { RefreshTokens.token eq refreshTokenValue }
                .singleOrNull()
        } ?: throw IllegalArgumentException("유효하지 않은 Refresh Token입니다.")

        // 만료 체크
        if (LocalDateTime.now().isAfter(tokenData[RefreshTokens.expiryDate])) {
            transaction {
                RefreshTokens.deleteWhere { RefreshTokens.token eq refreshTokenValue }
            }
            throw IllegalArgumentException("만료된 Refresh Token입니다.")
        }

        // 사용자 조회
        val userId = tokenData[RefreshTokens.userId].value
        val user = transaction {
            Users.selectAll().where { Users.id eq userId }
                .single()
        }

        val email = user[Users.email]

        // 새 토큰 생성
        val newToken = jwtConfig.generateToken(email)
        val newRefreshToken = jwtConfig.generateRefreshToken(email)

        // Refresh token 갱신
        transaction {
            RefreshTokens.deleteWhere { RefreshTokens.token eq refreshTokenValue }
        }
        saveRefreshToken(userId, newRefreshToken)

        val userDto = UserDto(
            id = user[Users.id].value,
            name = user[Users.name],
            email = user[Users.email],
            role = user[Users.role]
        )

        return AuthResponse(newToken, newRefreshToken, userDto)
    }

    fun getCurrentUser(email: String): UserDto {
        return transaction {
            Users.selectAll().where { Users.email eq email }
                .singleOrNull()
                ?.let {
                    UserDto(
                        id = it[Users.id].value,
                        name = it[Users.name],
                        email = it[Users.email],
                        role = it[Users.role]
                    )
                } ?: throw IllegalArgumentException("사용자를 찾을 수 없습니다.")
        }
    }

    fun logout(email: String) {
        transaction {
            val user = Users.selectAll().where { Users.email eq email }.singleOrNull()
            if (user != null) {
                RefreshTokens.deleteWhere { RefreshTokens.userId eq user[Users.id] }
            }
        }
    }

    private fun saveRefreshToken(userId: Long, tokenValue: String) {
        transaction {
            RefreshTokens.insert {
                it[token] = tokenValue
                it[RefreshTokens.userId] = userId
                it[expiryDate] = LocalDateTime.now().plusDays(7)
            }
        }
    }
}
